core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java

```diff
@@ -0,0 +1,121 @@
+                logger.info("[ARTIFACT-DOWNLOAD] [SUCCESS] [" + duration + "] [" + mavenCoordinate.getCoordinate() + "] [NONE] Artifact retrieved from repo: " + repos.get(i));
+                return jar.get();
+            } else if (found && i == repos.size() - 1) {
+                throw new MissingArtifactException("Artifact couldn't be retrieved for repo: " + repos.get(i), null);
+            } else if (found) {
+                continue;
+            }
+
+            for (var s : defaultPackaging) {
+                startTime = System.nanoTime();
+                try {
+                    found = true;
+                    jar = httpGetFile(mavenCoordinate.toProductUrl(repos.get(i), s));
+                } catch (MissingArtifactException e) {
+                    found = false;
+
+                    long duration = computeDurationInMs(startTime);
+                    logger.warn("[ARTIFACT-DOWNLOAD] [UNPROCESSED] [" + duration + "] [" + mavenCoordinate.getCoordinate() + "] [" + e.getClass().getSimpleName() + "] Artifact couldn't be retrieved for repo: " + repos.get(i), e);
+                }
+
+                if (jar.isPresent()) {long duration = computeDurationInMs(startTime);
+                    logger.info("[ARTIFACT-DOWNLOAD] [SUCCESS] [" + duration + "] [" + mavenCoordinate.getCoordinate() + "] [NONE] Artifact retrieved from repo: " + repos.get(i));
+                    return jar.get();
+                } else if (found && i == repos.size() - 1) {
+                    throw new MissingArtifactException("Artifact couldn't be retrieved for repo: " + repos.get(i), null);
+                } else if (found) {
+                    break;
+                }
+            }
+        }
+        throw new MissingArtifactException(
+                mavenCoordinate.toURL(mavenCoordinate.getMavenRepos().size() > 0
+                        ? mavenCoordinate.getMavenRepos().get(0)
+                        : "no repos specified") + " | "
+                        + mavenCoordinate.getPackaging(), null);
+    }
+
+    /**
+     * Utility function that stores the contents of GET request to a temporary file.
+     */
+    private static Optional<File> httpGetFile(final String url) throws MissingArtifactException {
+        Path tempFile = null;
+        try {
+            final var packaging = url.substring(url.lastIndexOf("."));
+            tempFile = Files.createTempFile("fasten", packaging);
+
+            final InputStream in = new URL(url).openStream();
+            Files.copy(in, tempFile, StandardCopyOption.REPLACE_EXISTING);
+            in.close();
+
+            return Optional.of(new File(tempFile.toAbsolutePath().toString()));
+        } catch (IOException e) {
+            if (tempFile != null) {tempFile.toFile().delete();}
+            throw new MissingArtifactException(e.getMessage(), e.getCause());
+        }
+    }
+
+    private long computeDurationInMs(long startTime) {
```

Comment:
Taking the time for the download should be removed, this is unrelated for the task at hand. If this is necessary for a specific case, I would solve it through implementing through subclassing and overriding the corresponding method (the override can then instruments it with timing logic).

Reply:
I am not sure whether we can remove the timing info there. Wouter changed the format of OPAL's logs and also, IIRC included timing for his monitoring solution.